magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala

```diff
@@ -152,28 +171,28 @@ object DefaultRecordPersistence extends Protocols with DiffsonProtocol with Reco
                          and data->${propertyWithLink.propertyName} ??| ARRAY[$ids])"""
       }
 
-      val excludeSelector = if (idsToExclude.isEmpty) None else Some(sqls"recordId not in (${idsToExclude})")
+      val excludeSelector = if (idsToExclude.isEmpty) None else Some(sqls"recordId not in ($idsToExclude)")
 
       val selectors = Seq(Some(SQLSyntax.join(dereferenceSelectors.toSeq, SQLSyntax.or)), excludeSelector)
 
-      this.getRecords(session, aspectIds, optionalAspectIds, None, None, None, dereference, selectors)
+      this.getRecords(session, tenantId, aspectIds, optionalAspectIds, None, None, None, dereference, selectors)
     }
   }
 
-  def getRecordAspectById(implicit session: DBSession, recordId: String, aspectId: String): Option[JsObject] = {
+  def getRecordAspectById(implicit session: DBSession, recordId: String, tenantId: BigInt, aspectId: String): Option[JsObject] = {
     sql"""select RecordAspects.aspectId as aspectId, name as aspectName, data from RecordAspects
-          inner join Aspects using (aspectId)
-          where RecordAspects.recordId=$recordId
-          and RecordAspects.aspectId=$aspectId"""
+          inner join Aspects using (aspectId, tenantId)
+          where (RecordAspects.aspectId, RecordAspects.recordId, RecordAspects.tenantId)=($aspectId, $recordId, $tenantId)"""
       .map(rowToAspect)
       .single.apply()
   }
 
   def getPageTokens(implicit session: DBSession,
+                    tenantId: BigInt,
                     aspectIds: Iterable[String],
                     limit: Option[Int] = None,
                     recordSelector: Iterable[Option[SQLSyntax]] = Iterable()): List[String] = {
-    val whereClauseParts = aspectIdsToWhereClause(aspectIds) ++ recordSelector
+    val whereClauseParts = aspectIdsToWhereClause(tenantId, aspectIds) ++ recordSelector
```

Comment:
fixed

Reply:
In the edge case where `aspectIds` is empty, this will do paging based on all records, not just the records associated with the given tenant.

This is a general problem with `aspectIdsToWhereClause`. When that are aspects, it filters records by tenantId for _every_ aspect, when it only needs to be done once. And when there are no aspects, it doesn't do the tenant filtering at all.

Reply:
Improved.